gluten-data/src/main/java/org/apache/gluten/columnarbatch/ColumnarBatches.java

```diff
@@ -186,23 +185,14 @@ private static ColumnarBatch load(BufferAllocator allocator, ColumnarBatch input
       ColumnarBatch output =
           ArrowAbiUtil.importToSparkColumnarBatch(allocator, arrowSchema, cArray);
 
-      // Follow gluten input's reference count. This might be optimized using
-      // automatic clean-up or once the extensibility of ColumnarBatch is enriched
+      // Loaded Arrow ColumnarBatch lifecycle is controlled by the caller. The GC can help clean it.
       IndicatorVector giv = (IndicatorVector) input.column(0);
-      ImplicitClass.ArrowColumnarBatchRetainer retainer =
-          new ImplicitClass.ArrowColumnarBatchRetainer(output);
-      for (long i = 0; i < (giv.refCnt() - 1); i++) {
-        retainer.retain();
-      }
-
       // close the input one
       for (long i = 0; i < giv.refCnt(); i++) {
         input.close();
       }
 
-      // populate new vectors to input
-      transferVectors(output, input);
-      return input;
+      return output;
```

Comment:
Thank you for writing the test and I finally got what happened here.

`transferVectors` was used for a reason. Once `ensureLoaded` is called, we should prevent caller code from using the input batch again after the method returns to do further operations (e.g., load it again, etc.) so the code behavior will be more predictable . So it's likely better not to remove it in the patch.

The root cause of the issue is actually we didn't update the column vectors used by `ColumnarBatchRow` in `transferVectors`. A proper way is to update them as well as `ColumnarBatch#columns`.

And I push a commit to the PR to untangle test code, please pull that one before next push. Thanks.

Reply:
`ColumnarBatchRow` is a protected member in `ColumnarBatch`, we cannot access it.
We close the `input` in `loaded` function, so if the caller revoke the input, it will fail.

Reply:
ok, I know, do you think we should use reflection to invoke `ColumnarBatchRow`?

Reply:
Further more, if we just need the caller don't use `input` again, we can set `input` to null. What's other concern I missed?

Reply:
Now the caller release input batch but not close output batch, to avoid change too much, I will try to transfer vector to ColumnarBatchRow by reflection

Reply:
> Further more, if we just need the caller don't use `input` again, we can set `input` to null. What's other concern I missed?

It's impossible... Java/Scala don't allow you to do that.

Reply:
> Now the caller release input batch but not close output batch, to avoid change too much, I will try to transfer vector to ColumnarBatchRow by reflection

Sounds good to me.